### ISSUE TITLE
Media queries always on top?

### ISSUE THREAD

**Issue by cretz** [japgolly/scalacss#47]
I am trying to mimic some CSS in the wild that looks like this:

```css
.main {
  padding: 20px;
}
@media (min-width: 768px) {
  .main {
    padding-right: 40px;
    padding-left: 40px;
  }
}
```

When in this order, browsers respect all values, but if the media queries are above, then they are not respected and they get overridden. So when I write a direct translation like:

```scala
  ".main" - (
    padding(20 px),
    &(media.minWidth(768 px)) - (
      paddingRight(40 px),
      paddingLeft(40 px)
    )
  )
```
It puts the media queries up top. I can't even break apart the rules to keep the media queries from going to the top. And even when I write this:

```scala
  ".main" - padding(20 px)

  "@media (min-width: 768px)" - (
    &(".main") - (
      paddingRight(40 px),
      paddingLeft(40 px)
    )
  )
```

I get:

```css
.main {
  padding: 20px;
}

@media (min-width: 768px) .main {
  padding-right: 40px;
  padding-left: 40px;
}
```

Which is not the same as having ``.main`` nested inside. Any ideas on how to achieve my goal?

**Comment by japgolly** [comment]
The implicit `Render[String]` in the Defaults does that.
https://github.com/japgolly/scalacss/blob/master/core/src/main/scala/scalacss/Renderer.scala#L26-L29

For now you could write your own implicit `Render[String]` that does that bit differently.

More permanently, do you think we should change the defaults to always put MQs at the bottom instead of the top?

**Comment by cretz** [comment]
I would suggest that the order they are in the code should be the determining factor.

**Comment by cretz** [comment]
Also, for the latter problem, how can I generate this:

```css
@media (min-width: 768px) {
  .main {
    padding-right: 40px;
    padding-left: 40px;
  }
}
```

I can't find a way to generate that (regardless of order). Sorry to couple problems into a single issue.

**Comment by japgolly** [comment]
> I would suggest that the order they are in the code should be the determining factor.

I don't think it's wise to create an environment where order is important. In fact it's a goal that I shouldn't have to care about or consider order when I create CSS with ScalaCss.

> how can I generate this:

standalone
```scala
  ".main" -
    &(media.minWidth(768 px))(
      padding.horizontal(40 px))
```

inline
```scala
  val inline = style("main")(
    &(media.minWidth(768 px))(
      padding.horizontal(40 px)
    )
  )
```

**Comment by cretz** [comment]
The problem is that, when building "cascading" stylesheets, order is important and it should be at the author's discretion. Granted if you are the author of the entire CSS file, you can make order less important, but as rules are combined together from disparate places around the app it might matter. An interchangeable renderer is probably good enough though.

Thanks for the snippet btw, it worked.

**Comment by japgolly** [comment]
Hmmm, right. Cascading in stylesheets is equivalent to inheritance and I believe the disadvantages & penalties of inheritance (over composition) have well been discussed and publicised. Philosophically, I'd really like to move myself and others away from that kind of CSS, and provide tools (like conflict detection) to do so.

But as you say, it should be at the author's discretion and that's something I want to be careful not to obstruct. One should be free to use this lib however they want.

So what to do? Well, I think it'd be better to keep the default as is, merging MQs  because it's pedagogic recommendation and it produces smaller CSS. We can add an order sensitive renderer and make it opt-in. Or maybe we could make order-sensitivity the default for standalone CSS and leave the inline default as is?

> Thanks for the snippet btw, it worked.
sweet :+1:

**Comment by cretz** [comment]
For anyone else needing this, here's a simple renderer (but you can't have ``Defaults._`` in scope due to ambiguous implicits):

```scala
implicit val inOrderRenderer = new Renderer[String] {
  override def apply(css: Css): String = {
    val sb = new StringBuilder
    val fmt = StringRenderer.formatPretty()(sb)
    css.foreach { case CssEntry(mq, sel, content) =>
      mq.foreach(fmt.mqStart)
      fmt.selStart(mq, sel)
      fmt.kv1(mq, content.head)
      content.tail.foreach(fmt.kvn(mq, _))
      fmt.selEnd(mq, sel)
      mq.foreach(fmt.mqEnd)
    }
    fmt.done()
    sb.toString()
  }
}
implicit val env = Env.empty
```
The code in the repo: https://github.com/cretz/scala-web-ideal/blob/1f45cb3d71f0f0690b6b551a8d1de16f45263c57/server/src/main/scala/webideal/util/ScalaCssSupport.scala#L18

**Comment by japgolly** [comment]
@cretz I'd recommend creating your own settings object extending one of the `DevDefaults`, `ProdDefaults`, `IDefaults` traits, then overriding the string renderer with your own.

```scala
object MyCssSettings extends DevDefaults {
  override implicit def stringRenderer: Renderer[String] = ???
}

// Then instead of import Defaults._
import MyCssSettings._
```

**Comment by japgolly** [comment]
Closed via 6585dbf418fcccc43e822992b2c61d0f779409a2.

You can use [`StringRenderer.Raw`](https://github.com/japgolly/scalacss/blob/6585dbf418fcccc43e822992b2c61d0f779409a2/core/src/main/scala/scalacss/Renderer.scala#L32-L44).
